dhkem/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+//! RFC returns the encapsulated key and an extracted shared secret, while our
+//! implementation leaves the extraction process up to the user. This type of KEM
+//! construction is currently being used in HPKE, as per the RFC, and in the current
+//! draft of the [TLS KEM
+//! combiner](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-10).
+
+use kem::{Decapsulate, Encapsulate};
+use rand_core::CryptoRngCore;
+#[cfg(feature = "zeroize")]
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// Newtype for a piece of data that may be encapsulated
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Default)]
+pub struct DhEncapsulator<X>(X);
+/// Newtype for a piece of data that may be decapsulated
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Default)]
+pub struct DhDecapsulator<X>(X);
+
+impl<X> AsRef<X> for DhEncapsulator<X> {
+    fn as_ref(&self) -> &X {
+        &self.0
+    }
+}
+
+impl<X> From<X> for DhEncapsulator<X> {
+    fn from(value: X) -> Self {
+        Self(value)
+    }
+}
+
+impl<X> AsRef<X> for DhDecapsulator<X> {
+    fn as_ref(&self) -> &X {
+        &self.0
+    }
+}
+
+impl<X> From<X> for DhDecapsulator<X> {
+    fn from(value: X) -> Self {
+        Self(value)
+    }
+}
+
+impl<X> DhEncapsulator<X> {
+    /// Consumes `self` and returns the wrapped value
+    pub fn into_inner(self) -> X {
+        self.0
+    }
+}
+
+impl<X> DhDecapsulator<X> {
+    /// Consumes `self` and returns the wrapped value
+    pub fn into_inner(self) -> X {
+        self.0
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<X: Zeroize> Zeroize for DhEncapsulator<X> {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<X: Zeroize> Zeroize for DhDecapsulator<X> {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<X: ZeroizeOnDrop> ZeroizeOnDrop for DhEncapsulator<X> {}
+
+#[cfg(feature = "zeroize")]
+impl<X: ZeroizeOnDrop> ZeroizeOnDrop for DhDecapsulator<X> {}
+
+/// This is a trait that all KEM models should implement, and should probably be
+/// promoted to the kem crate itself. It specifies the types of encapsulating and
+/// decapsulating keys created by key generation, the shared secret type, and the
+/// encapsulated key type
+pub trait DhKem {
+    /// The type that will implement [`Decapsulate`]
+    type DecapsulatingKey: Decapsulate<Self::EncapsulatedKey, Self::SharedSecret>;
+
+    /// The type that will implement [`Encapsulate`]
+    type EncapsulatingKey: Encapsulate<Self::EncapsulatedKey, Self::SharedSecret>;
+
+    /// The type of the encapsulated key
+    type EncapsulatedKey;
+
+    /// The type of the shared secret
+    type SharedSecret;
+
+    /// Generates a new (decapsulating key, encapsulating key) keypair for the KEM
+    /// model
+    fn random_keypair(
+        rng: &mut impl CryptoRngCore,
+    ) -> (Self::DecapsulatingKey, Self::EncapsulatingKey);
+}
```

Comment:
This seems somewhat similar to the `EncappedKey` trait from `kem` v0.2 which was removed in the latest prerelease: https://docs.rs/kem/0.2.0/kem/trait.EncappedKey.html

cc @rozbb